### PR TITLE
Harden JIT job control against cctors deadlocking.

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

File: mono/mini/mini-runtime.c [hunk 1]
```diff
@@ -1983,9 +1990,14 @@ wait_or_register_method_to_compile (MonoMethod *method, MonoDomain *domain)
 
 		unlock_compilation_data ();
 		return FALSE;
-	} else if (jit_tls->active_jit_methods > 0) {
+	} else if (jit_tls->active_jit_methods > 0 || mono_threads_is_current_thread_in_protected_block ()) {
 		//We can't suspend the current thread if it's already JITing a method.
 		//Dependency management is too compilated and we want to get rid of this anyways.
+
+		//We can't suspend the current thread if it's running a protected block (such as a cctor)
+		//We can't relie only on JIT nesting as cctor's can be run from outside the JIT.
```
Comment (luhenry) [inline mono/mini/mini-runtime.c]: `relie` -> `rely`

File: mono/mini/mini-runtime.c [hunk 2]
```diff
@@ -1983,9 +1990,14 @@ wait_or_register_method_to_compile (MonoMethod *method, MonoDomain *domain)
...
+
+		//Finally, he hit a timeout or spurious wakeup. We're better of just giving up and keep recompiling
```
Comment (jaykrell) [inline mono/mini/mini-runtime.c]: of => off

**Comment by kumpera** [comment]
Fixed spelling.

File: mono/mini/mini-runtime.c [hunk 3]
```diff
@@ -2008,15 +2017,25 @@ wait_or_register_method_to_compile (MonoMethod *method, MonoDomain *domain)
 			++entry->threads_waiting;
 
 			g_assert (entry->has_cond);
-			mono_coop_cond_wait (&entry->cond, &compilation_data.lock);
+			mono_coop_cond_timedwait (&entry->cond, &compilation_data.lock, MAX_JIT_TIMEOUT_MS);
 			--entry->threads_waiting;
```
Comment (vargaz) [inline mono/mini/mini-runtime.c]: Doesn't this going to mask bugs ?

Comment (kumpera) [reply mono/mini/mini-runtime.c]: Yes, it will.
OTH, it ensures progress can be made even if slow enough. Do you think it's not worth it?

**Comment by luhenry** [comment]
@monojenkins build Windows x64

**Comment by luhenry** [comment]
@monojenkins build Windows i386

**Comment by luhenry** [comment]
@monojenkins build OS X x64

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Acceptance Tests

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64

**Comment by luhenry** [comment]
@marek-safar I am not sure we wanted to merge it :wink:

@kumpera can you confirm if we should revert or keep it?

**Comment by kumpera** [comment]
I think me and Zoltan were undecided on whether use a timeout and ensure progress or keep it invite and make bugs evident.

Merging this is fine. In case we decide of the later, I'll do a followup PR.

**Comment by kumpera** [comment]
@vargaz what do you think?

**Comment by vargaz** [comment]
I didn't like the timeout, but its fine.

**Pull request merged**